protocol/xmrtaker/swap_state.go

```diff
@@ -317,22 +318,51 @@ func (s *swapState) doRefund() (ethcommon.Hash, error) {
 }
 
 func (s *swapState) tryRefund() (ethcommon.Hash, error) {
-	untilT0 := time.Until(s.t0)
-	untilT1 := time.Until(s.t1)
+	stage, err := s.Contract().Swaps(s.CallOpts(), s.contractSwapID)
+	if err != nil {
+		return ethcommon.Hash{}, err
+	}
+	switch stage {
+	case swapfactory.StageInvalid:
+		return ethcommon.Hash{}, errRefundInvalid
+	case swapfactory.StageCompleted:
+		return ethcommon.Hash{}, errRefundSwapCompleted
+	case swapfactory.StagePending, swapfactory.StageReady:
+		// do nothing
+	default:
+		panic("Unhandled stage value")
+	}
+	isReady := stage == swapfactory.StageReady
 
-	isReady, err := s.Contract().IsReady(s.CallOpts(), s.contractSwapID)
+	ts, err := s.LatestBlockTimestamp(s.ctx)
 	if err != nil {
 		return ethcommon.Hash{}, err
 	}
 
-	log.Debugf("tryRefund isReady=%v untilT0=%vs untilT1=%vs", isReady, untilT0.Seconds(), untilT1.Seconds())
+	log.Debugf("tryRefund isReady=%v untilT0=%vs untilT1=%vs",
+		isReady, ts.Sub(s.t0).Seconds(), ts.Sub(s.t1).Seconds())
 
-	if (untilT0 > 0 && isReady) && untilT1 > 0 {
-		// we've passed t0 but aren't past t1 yet, so we need to wait until t1
-		log.Infof("waiting until time %s to refund", s.t1)
-		<-time.After(untilT1)
+	if ts.Before(s.t0) && !isReady {
+		// using errT0 to avoid shadow warnings
+		txHash, errT0 := s.refund()
```

Comment:
```suggestion
		txHash, err := s.refund() //nolint:govet
```

i *think* govet is the linter that complains? either way, i'd just throw a nolint in there